release_dashboard/lib/widgets/push_release_substep.dart

```diff
@@ -0,0 +1,113 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:conductor_core/conductor_core.dart';
+import 'package:flutter/material.dart';
+import 'package:provider/provider.dart';
+
+import '../models/conductor_status.dart';
+import '../state/status_state.dart';
+import 'common/checkbox_substep.dart';
+import 'common/url_button.dart';
+
+enum PushReleaseSubstep {
+  pushRelease,
+  verifyRelease,
+}
+
+/// Group and display all substeps of the push release step.
+///
+/// This is the last step of the release.
+class PushReleaseSubsteps extends StatefulWidget {
+  const PushReleaseSubsteps({
+    Key? key,
+    required this.nextStep,
+  }) : super(key: key);
+
+  final VoidCallback nextStep;
+
+  @override
+  State<PushReleaseSubsteps> createState() => PushReleaseSubstepsState();
+
+  static const Map<PushReleaseSubstep, String> substepTitles = <PushReleaseSubstep, String>{
+    PushReleaseSubstep.pushRelease: 'Push to the release channel branch',
+    PushReleaseSubstep.verifyRelease: 'Verify the release',
+  };
+  static const Map<PushReleaseSubstep, String> substepSubtitles = <PushReleaseSubstep, String>{
+    PushReleaseSubstep.pushRelease: 'A special access is needed to push to the',
+    PushReleaseSubstep.verifyRelease: 'After the push, please make sure the release passes all packaging builds here: ',
+  };
+}
+
+class PushReleaseSubstepsState extends State<PushReleaseSubsteps> {
+  final Map<PushReleaseSubstep, bool> _isEachSubstepChecked = <PushReleaseSubstep, bool>{};
+
+  @override
+  void initState() {
+    /// If [substep] is false, that substep is unchecked, otherwise, it is checked.
+    ///
+    /// All substeps are unchecked at the beginning.
+    for (final PushReleaseSubstep substep in PushReleaseSubstep.values) {
+      _isEachSubstepChecked[substep] = false;
+    }
+    super.initState();
+  }
+
+  /// Toggle the boolean value [substepName].
+  void substepPressed(PushReleaseSubstep substep) {
+    setState(() {
+      _isEachSubstepChecked[substep] = !_isEachSubstepChecked[substep]!;
+    });
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    Map<ConductorStatusEntry, Object>? releaseStatus = context.watch<StatusState>().releaseStatus;
+    final String postMonitorLuci = luciConsoleLink(
+      releaseStatus?[ConductorStatusEntry.releaseChannel] as String? ?? 'master',
+      'packaging',
+    );
+
+    return Column(
+      crossAxisAlignment: CrossAxisAlignment.center,
+      children: <Widget>[
+        CheckboxAsSubstep(
+          substepName: PushReleaseSubsteps.substepTitles[PushReleaseSubstep.pushRelease]!,
+          subtitle: SelectableText('${PushReleaseSubsteps.substepSubtitles[PushReleaseSubstep.pushRelease]!} '
+              '${releaseStatus?[ConductorStatusEntry.releaseChannel]} branch directly. '
```

Comment:
can this be null at this point?

Reply:
> No, but when the whole app gets loaded in tests, it cases a null check failure. They are only here for the tests to pass.

Can we provide a `releaseStatus` in the tests?

Reply:
In `conductor_status.dart`, there are 2 tests where the app doesn't have any `releaseStatus` (to simulate the situation where a release has not been initialized yet). And these tests would fail exactly right here indicating that a null check has been performed on a null value.
Here is the test: https://github.com/Yugue/cocoon/blob/0f9fd2fdd379bd5ec6b487d6248b46e354fa531e/release_dashboard/test/widgets/conductor_status_test.dart#L22

Reply:
Why would that test render `PushReleaseSubsteps`?

Reply:
It doesn't directly. But I think `Stepper` paints all the widgets. It displays `PushReleaseSubsteps` although it is not at that step yet. Even if the tests in `conductor_status_test.dart` were changed, the test in `main_test.dart` still fails because it all the entire stepper with a null state [here](https://github.com/Yugue/cocoon/blob/d7623bedb388aafa1535b2c21ec0b09fb673264e/release_dashboard/test/main_test.dart#L16). A a null check error would throw exactly at this line if it were changed to `!`. Therefore, `?` was used to not crash the tests when the entire stepper was painted with no state (which could happen in a scenario where a release hasn't been initialized yet).

Reply:
> But it seems that Stepper paints all the widgets in each step. It renders PushReleaseSubsteps although it is not at that step yet.
Ahh, I see.

I think the solution here would be to conditionally render the `MainProgression` widget only when there is a existing state, rather than coding the `MainProgression` and its subwidgets to handle a null state when that doesn't really make sense.

Reply:
I don't think I can conditionally render `MainProgression`. Since when the state is null (a scenario of no release started yet), the user should still be able to access the first step of `MainProgression`, which is to create a new release.

I can see if I can implement a logic to only paint the current step's widget when we are at the current step.

Reply:
Oh, good point

Reply:
I added a todo [here](https://github.com/Yugue/cocoon/blob/1aa19c0752d3f7bd34a37548591ea6a0f2d5edf5/release_dashboard/lib/widgets/progression.dart#L90). 

It will be handled in this [PR](https://github.com/flutter/cocoon/pull/1500).